node/debug.go

```diff
@@ -0,0 +1,80 @@
+package node
+
+import (
+	"net"
+
+	"github.com/tendermint/tendermint/config"
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/libs/service"
+	rpccore "github.com/tendermint/tendermint/rpc/core"
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/store"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Debug is a type useful for debugging tendermint problems.
+// Tendermint nodes will shutdown if a divergent hash is detected. Once in this
+// state, they will not start up again. Debug runs just an RPC server on the
+// tendermint data stores without running any other components. This way a user
+// can query the RPC server to diagnose the issue that caused a crash to begin with.
+type Debug struct {
```

Comment:
I wouldn't export this, and just return a service.Service from the constructors. 

also, is this *actually* a thing that we want to have be part of the public interface of `node` or would this be better set in some kind of testutil package or similar, or do we expect to expose it via a CLI? I think not exporting it from the package and just providing a service.Service constructor is right either way.

Reply:
Re: Return value. As a rule, I advise returning _concrete_ types rather than interfaces, except in cases where you can't (e.g., a factory that returns different implementations of an interface depending on its parameters). https://github.com/golang/go/wiki/CodeReviewComments#interfaces discusses this a bit. Returning interfaces "hides" the concrete type, but makes it harder to update the interface later.

Re: Should we export this. This is a question I'm also interested in—it feels like something we should provide via an ancillary package.

Reply:
I agree that this should be private and the constructors should return the service.Service interface in keeping with the other node implementations. I do get your point @creachadair, but I do think a Service which has a very specific and simple interface that is unlikely to change is likely an exception to that rule.

Reply:
I would also want this to be integrated with the CLI (by adding Debug to `newDefaultNode()` function). Debug should be a mode type. Just like seed, validator and full. A user should be able to run
```
tendermint start --mode debug 
```
to start their existing node in debug mode

Reply:
The intent of that recommendation is not about the complexity of the interface, but the impact on future maintenance. Returning an interface commits you to updating everything atomically when the interface changes (even "safe" extension changes that otherwise consumers wouldn't care about).

That said, it's not going to result in incorrect code to do it that way.

Reply:
Latest change moves to be somewhat simpler than `service.Service`. Resolving this comment for now.